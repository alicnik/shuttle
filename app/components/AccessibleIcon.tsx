import * as AccessibleIconComponent from '@radix-ui/react-accessible-icon';

export const AccessibleIcon = AccessibleIconComponent.Root;
