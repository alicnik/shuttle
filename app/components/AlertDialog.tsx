import {
  AlertDialog as AlertDialogComponent,
  AlertDialogAction,
  AlertDialogCancel,
  AlertDialogContent,
  AlertDialogDescription,
  AlertDialogFooter,
  AlertDialogHeader,
  AlertDialogTitle,
} from '~/components/ui/alert-dialog';

type AlertDialogProps = React.PropsWithChildren<{
  title: string;
  description: string;
  onConfirm: () => void;
}>;

export function AlertDialog({
  children,
  title,
  description,
  onConfirm,
}: AlertDialogProps) {
  return (
    <AlertDialogComponent>
      {children}
      <AlertDialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <AlertDialogHeader>
          <AlertDialogTitle>{title}</AlertDialogTitle>
          <AlertDialogDescription>{description}</AlertDialogDescription>
        </AlertDialogHeader>
        <AlertDialogFooter>
          <AlertDialogCancel>Cancel</AlertDialogCancel>
          <AlertDialogAction onClick={onConfirm}>Continue</AlertDialogAction>
        </AlertDialogFooter>
      </AlertDialogContent>
    </AlertDialogComponent>
  );
}
