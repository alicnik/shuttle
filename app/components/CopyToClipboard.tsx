import { ClipboardCopyIcon } from '@radix-ui/react-icons';
import { useToast } from './ui';
import { Tooltip } from './Tooltip';
import type { TooltipProps } from './Tooltip';

interface CopyToClipboardProps {
  copyText: string;
  successMessage: string;
  tooltipOptions: Omit<TooltipProps, 'children'>;
}

export function CopyToClipboard({
  copyText,
  successMessage,
  tooltipOptions: { content, side = 'top', sideOffset = 0 },
}: CopyToClipboardProps) {
  const { toast } = useToast();

  return (
    <Tooltip content={content} side={side} sideOffset={sideOffset}>
      <ClipboardCopyIcon
        className="translate-y-[3px] cursor-pointer"
        onClick={async () => {
          await navigator.clipboard.writeText(copyText);
          toast({ description: successMessage });
        }}
      />
    </Tooltip>
  );
}
