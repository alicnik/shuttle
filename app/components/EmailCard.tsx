import type { Email } from '@prisma/client';
import { Checkbox } from '~/components/ui/checkbox';
import { useNavigation, useSubmit } from '@remix-run/react';
import { formatRelative } from 'date-fns';

interface EmailCardProps {
  email: Email;
  selected: string[];
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
  preview: string | null;
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
}

export function EmailCard({
  email,
  selected,
  setSelected,
  preview,
  setPreview,
}: EmailCardProps) {
  const submit = useSubmit();
  const navigation = useNavigation();

  if (navigation.state === 'submitting') {
    console.log(Object.fromEntries(navigation.formData || []));
  }

  const isOptimisticRead =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markRead' &&
    navigation.formData?.get('emailId') === email.id;

  const isOptimisticUnread =
    ['submitting', 'loading'].includes(navigation.state) &&
    navigation.formData?.get('_action') === 'markSelectedUnread' &&
    String(navigation.formData?.get('selected')).split(',').includes(email.id);

  const isSelected = selected.includes(email.id);
  const isPreview = preview === email.id;

  const from = email.from.startsWith('<')
    ? email.from.slice(1, -1)
    : email.from.split('<')[0].trim();

  return (
    <div
      data-selected={isSelected}
      data-in-view={isPreview}
      className="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-800 data-[in-view=true]:bg-zinc-800 data-[selected=true]:bg-zinc-800"
    >
      <Checkbox
        className="mt-[2px]"
        checked={isSelected || isPreview}
        onCheckedChange={(checked) => {
          setSelected((selected) => {
            if (checked) {
              return [...selected, email.id];
            }

            if (isPreview) {
              setPreview(null);
            }
            return selected.filter((id) => id !== email.id);
          });
        }}
      />
      <div
        data-read={!isOptimisticUnread && (email.read || isOptimisticRead)}
        className="cursor-pointer text-left data-[read=true]:opacity-50"
        onClick={() =>
          setPreview((current) => {
            if (current) {
              submit(
                { emailId: current, _action: 'markRead' },
                { method: 'post' }
              );
            }
            return email.id;
          })
        }
      >
        <div className="mb-1 flex items-center ">
          <p className="mr-12 w-max text-sm font-bold">{from}</p>
          <p className="ml-auto w-max text-xs">
            {formatRelative(email.createdAt, new Date())}
          </p>
        </div>
        <p className="mb-1 text-sm">{email.subject}</p>
        <p className="line-clamp-2 text-xs">{email.text}</p>
      </div>
    </div>
  );
}
