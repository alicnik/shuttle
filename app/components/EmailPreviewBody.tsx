import clsx from 'clsx';
import type { Email } from '@prisma/client';
import * as React from 'react';

export function EmailPreviewBody({ html }: { html?: Email['html'] }) {
  const htmlWithBaseTag = React.useMemo(() => {
    if (!html) return '';

    // This ensures links open in a new tab and not in the iframe
    const baseTag = '<base target="_blank">';
    const headTag = '<head>';
    const headTagIndex = html.indexOf(headTag);

    if (headTagIndex === -1) {
      return html;
    }

    return (
      html.slice(0, headTagIndex + headTag.length) +
      baseTag +
      html.slice(headTagIndex + headTag.length)
    );
  }, [html]);

  return (
    <iframe
      data-testid="email-preview-body"
      className={clsx('email-preview', 'h-full w-full')}
      srcDoc={htmlWithBaseTag}
      title="Email preview"
    />
  );
}
