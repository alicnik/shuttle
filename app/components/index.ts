export * from './AlertDialog';
export * from './EmailCard';
export * from './EmailPreviewBody';
export * from './EmailPreviewHeader';
export * from './Tooltip';
