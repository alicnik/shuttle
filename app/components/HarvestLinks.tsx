import * as React from 'react';
import {
  Button,
  Dialog,
  DialogContent,
  DialogDescription,
  DialogHeader,
  DialogTitle,
  DialogTrigger,
  useToast,
} from './ui';
import { Tooltip } from './Tooltip';
import { Link2Icon } from '@radix-ui/react-icons';
import { LINK_COPY_SUCCESS_MESSAGE } from '~/lib';
import type { Email } from '@prisma/client';
import { AccessibleIcon } from './AccessibleIcon';

interface HarvestLinksProps {
  email: Email;
}

export function HarvestLinks({ email }: HarvestLinksProps) {
  const { toast } = useToast();

  const links = React.useMemo(() => {
    /* c8 ignore next 3 */
    if (typeof window === 'undefined') {
      return [];
    }
    const parser = new DOMParser();
    const doc = parser.parseFromString(email.html, 'text/html');
    const links = Array.from(doc.querySelectorAll('a'));
    const mapped = links.map((link) => {
      const linkDoc = parser.parseFromString(link.innerHTML, 'text/html');
      const elements = linkDoc.querySelectorAll('*');
      elements.forEach((element) => {
        element.removeAttribute('style');
        element.removeAttribute('width');
        element.removeAttribute('height');
        if (element.tagName === 'IMG') {
          element.setAttribute('style', 'max-width: 80%;');
        }
      });
      link.innerHTML = linkDoc.body.innerHTML;
      return link;
    });
    return mapped;
  }, [email.html]);

  return (
    <Dialog>
      <Tooltip content="Harvest links">
        <DialogTrigger asChild>
          <Button
            variant="ghost"
            className="hover:bg-zinc-300"
            type="button"
            size="icon"
          >
            <AccessibleIcon label="Harvest links from email">
              <Link2Icon />
            </AccessibleIcon>
          </Button>
        </DialogTrigger>
      </Tooltip>
      <DialogContent onCloseAutoFocus={(e) => e.preventDefault()}>
        <DialogHeader>
          <DialogTitle>Found the following links in the email</DialogTitle>
          <DialogDescription>
            Click a link to copy it to your clipboard
          </DialogDescription>
          {links.length > 0 ? (
            <ul className="grid auto-rows-fr grid-cols-2 place-items-center gap-x-2 gap-y-8 pt-8">
              {links.map((link) => {
                return (
                  <Tooltip key={link.href} content={link.href} sideOffset={4}>
                    <li
                      className="flex cursor-pointer justify-center"
                      dangerouslySetInnerHTML={{
                        __html: link.innerHTML,
                      }}
                      onClick={async () => {
                        await navigator.clipboard.writeText(link.href);
                        toast({
                          description: LINK_COPY_SUCCESS_MESSAGE,
                        });
                      }}
                    />
                  </Tooltip>
                );
              })}
            </ul>
          ) : (
            <DialogDescription>
              No links were found in the email
            </DialogDescription>
          )}
        </DialogHeader>
      </DialogContent>
    </Dialog>
  );
}
