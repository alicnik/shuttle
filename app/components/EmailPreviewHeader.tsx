import { Form, Link, useSubmit } from '@remix-run/react';
import { format } from 'date-fns';
import { Tooltip } from './Tooltip';
import { Button, Separator } from './ui';
import type { Email, User } from '@prisma/client';
import {
  EnvelopeClosedIcon,
  ExternalLinkIcon,
  TrashIcon,
} from '@radix-ui/react-icons';
import { AlertDialog } from './AlertDialog';
import { AlertDialogTrigger } from './ui/alert-dialog';

interface EmailPreviewHeaderProps {
  email: Email;
  userId: User['id'];
  setPreview: React.Dispatch<React.SetStateAction<string | null>>;
  setSelected: React.Dispatch<React.SetStateAction<string[]>>;
}

export function EmailPreviewHeader({
  email,
  userId,
  setPreview,
  setSelected,
}: EmailPreviewHeaderProps) {
  const submit = useSubmit();

  return (
    <div className="max-h-48 overflow-hidden p-4 text-zinc-900">
      <h3 className="truncate">
        <span className="font-bold">From:</span> {email.from}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">To:</span> {`${email.userId}@shuttle.email`}
      </h3>
      <h3 className="truncate">
        <span className="font-bold">Subject:</span> {email.subject}
      </h3>
      <h3 className="mb-4 truncate">
        <span className="font-bold">Received:</span>{' '}
        {format(email.createdAt, 'cccc, io MMMM yyyy @ hh:mm a')}
      </h3>
      <Form method="post" className="mb-4 flex gap-2">
        <input type="hidden" name="selected" value={email.id} />
        <Tooltip content="Mark as unread">
          <Button
            variant="ghost"
            size="icon"
            name="_action"
            value="markSelectedUnread"
          >
            <EnvelopeClosedIcon />
          </Button>
        </Tooltip>
        <AlertDialog
          title="Are you sure?"
          description="This action cannot be undone."
          onConfirm={() => {
            setPreview(null);
            setSelected((selected) => selected.filter((id) => id !== email.id));
            submit(
              { selected: email.id, _action: 'deleteSelected' },
              { method: 'post' }
            );
          }}
        >
          <Tooltip content="Delete">
            <AlertDialogTrigger asChild>
              <Button
                name="_action"
                value="deleteSelected"
                variant="ghost"
                size="icon"
                className="hover:bg-destructive focus:bg-destructive"
              >
                <TrashIcon />
              </Button>
            </AlertDialogTrigger>
          </Tooltip>
        </AlertDialog>
        <Tooltip content="Open in new window">
          <Link to={`/${userId}/${email.id}`} target="_blank">
            <Button
              type="button"
              variant="ghost"
              size="icon"
              onClick={() => {
                submit(
                  { emailId: email.id, _action: 'markRead' },
                  { method: 'post' }
                );
              }}
            >
              <ExternalLinkIcon />
            </Button>
          </Link>
        </Tooltip>
      </Form>
      <Separator />
    </div>
  );
}
