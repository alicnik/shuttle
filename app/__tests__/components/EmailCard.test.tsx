import {
  createMockFormData,
  setWindowWidth,
  setupRemixStub,
} from '../test-utils';
import { screen } from '@testing-library/react';
import { EmailCard } from '~/components';
import { useNavigation, useSearchParams } from '@remix-run/react';
import { fromPartial } from '@total-typescript/shoehorn';

vi.mock('@remix-run/react', async () => {
  const actual = (await vi.importActual(
    '@remix-run/react'
    // eslint-disable-next-line @typescript-eslint/consistent-type-imports
  )) as typeof import('@remix-run/react');
  return {
    ...actual,
    useSearchParams: vi.fn(() => [new URLSearchParams(), vi.fn()]),
    useNavigation: vi.fn(actual.useNavigation),
  };
});

const props = {
  email: {
    id: 'mock-id',
    subject: 'mock-subject',
    from: 'mock-from',
    text: 'mock-text',
    html: 'mock-html',
    userId: 'mock-user-id',
    read: false,
    createdAt: new Date('October 13, 2023 13:07:00'),
  },
  selected: [],
  setSelected: vi.fn(),
};

describe('EmailCard', () => {
  beforeEach(() => {
    setWindowWidth(1024);
  });

  it('should match the snapshot for mobile view', () => {
    setWindowWidth(375);

    const { container } = setupRemixStub(<EmailCard {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot for tablet view', () => {
    setWindowWidth(768);

    const { container } = setupRemixStub(<EmailCard {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('should match the snapshot for desktop view', () => {
    setWindowWidth(1024);

    const { container } = setupRemixStub(<EmailCard {...props} />);

    expect(container).toMatchSnapshot();
  });

  it('applies the data-selected attribute when the email is selected', () => {
    setupRemixStub(<EmailCard {...props} selected={['mock-id']} />);

    const card = screen.getByTestId('email-card');
    const checkbox = screen.getByRole('checkbox');

    expect(card).toHaveAttribute('data-selected', 'true');
    expect(checkbox).toBeChecked();
  });

  it('applies the data-in-view attribute when the email is being previewed', () => {
    vi.mocked(useSearchParams).mockReturnValueOnce([
      new URLSearchParams({
        preview: props.email.id,
      }),
      vi.fn(),
    ]);

    setupRemixStub(<EmailCard {...props} />);

    const card = screen.getByTestId('email-card');

    expect(card).toHaveAttribute('data-in-view', 'true');
  });

  it.each([
    { state: 'read', read: true },
    { state: 'unread', read: false },
  ])(`applies the data-read attribute when the email is $state`, ({ read }) => {
    setupRemixStub(<EmailCard {...props} email={{ ...props.email, read }} />);

    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('data-read', String(read));
  });

  it('applies the data-read attribute optimistically when the email is marked read', () => {
    const formData = createMockFormData({
      _action: 'markRead',
      emailId: props.email.id,
    });

    vi.mocked(useNavigation).mockReturnValueOnce(
      fromPartial({
        state: 'loading',
        formData,
      })
    );

    setupRemixStub(<EmailCard {...props} />);

    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('data-read', 'true');
  });

  it('disapplies the data-read attribute optimistically when the email is marked unread', () => {
    const formData = createMockFormData({
      _action: 'markSelectedUnread',
      selected: props.email.id,
    });

    vi.mocked(useNavigation).mockReturnValueOnce(
      fromPartial({
        state: 'loading',
        formData,
      })
    );

    setupRemixStub(
      <EmailCard {...props} email={{ ...props.email, read: true }} />
    );

    const button = screen.getByRole('button');

    expect(button).toHaveAttribute('data-read', 'false');
  });

  it('resets the selected emails when the card is clicked', async () => {
    const { user } = setupRemixStub(<EmailCard {...props} />);

    await user.click(screen.getByRole('button'));

    expect(props.setSelected).toHaveBeenCalledWith([]);
  });
});
