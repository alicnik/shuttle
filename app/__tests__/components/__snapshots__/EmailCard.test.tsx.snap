// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EmailCard > should match the snapshot for desktop view 1`] = `
<div>
  <div
    class="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-200 data-[in-view=true]:bg-zinc-200 data-[selected=true]:bg-zinc-200 dark:hover:bg-zinc-800 dark:data-[in-view=true]:bg-zinc-800 dark:data-[selected=true]:bg-zinc-800"
    data-in-view="false"
    data-selected="false"
    data-testid="email-card"
  >
    <button
      aria-checked="false"
      class="peer h-4 w-4 shrink-0 rounded-sm border border-primary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mt-[2px]"
      data-state="unchecked"
      role="checkbox"
      type="button"
      value="on"
    />
    <form
      action="/"
      class="w-full"
      method="post"
    >
      <input
        name="emailId"
        type="hidden"
        value="mock-id"
      />
       
      <button
        class="w-full cursor-pointer text-left data-[read=true]:opacity-50"
        data-read="false"
        name="_action"
        type="submit"
        value="setPreview"
      >
        <div
          class="mb-1 flex items-center justify-between "
        >
          <p
            class="w-max max-w-[180px] truncate text-sm font-bold"
          >
            mock-from
          </p>
          <p
            class="ml-auto w-max text-xs"
          >
            13:07
          </p>
        </div>
        <p
          class="mb-1 max-w-[250px] truncate text-sm"
        >
          mock-subject
        </p>
        <p
          class="line-clamp-2 max-w-[300px] text-xs"
        >
          mock-text
        </p>
      </button>
    </form>
  </div>
</div>
`;

exports[`EmailCard > should match the snapshot for mobile view 1`] = `
<div>
  <div
    class="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-200 data-[in-view=true]:bg-zinc-200 data-[selected=true]:bg-zinc-200 dark:hover:bg-zinc-800 dark:data-[in-view=true]:bg-zinc-800 dark:data-[selected=true]:bg-zinc-800"
    data-in-view="false"
    data-selected="false"
    data-testid="email-card"
  >
    <button
      aria-checked="false"
      class="peer h-4 w-4 shrink-0 rounded-sm border border-primary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mt-[2px]"
      data-state="unchecked"
      role="checkbox"
      type="button"
      value="on"
    />
    <a
      class="w-full"
      href="/mock-user-id/mock-id"
    >
      <button
        class="w-full cursor-pointer text-left data-[read=true]:opacity-50"
        data-read="false"
      >
        <div
          class="mb-1 flex items-center justify-between "
        >
          <p
            class="w-max max-w-[180px] truncate text-sm font-bold"
          >
            mock-from
          </p>
          <p
            class="ml-auto w-max text-xs"
          >
            13:07
          </p>
        </div>
        <p
          class="mb-1 max-w-[250px] truncate text-sm"
        >
          mock-subject
        </p>
        <p
          class="line-clamp-2 max-w-[300px] text-xs"
        >
          mock-text
        </p>
      </button>
    </a>
  </div>
</div>
`;

exports[`EmailCard > should match the snapshot for tablet view 1`] = `
<div>
  <div
    class="mb-4 flex gap-2 rounded-md p-2 hover:bg-zinc-200 data-[in-view=true]:bg-zinc-200 data-[selected=true]:bg-zinc-200 dark:hover:bg-zinc-800 dark:data-[in-view=true]:bg-zinc-800 dark:data-[selected=true]:bg-zinc-800"
    data-in-view="false"
    data-selected="false"
    data-testid="email-card"
  >
    <button
      aria-checked="false"
      class="peer h-4 w-4 shrink-0 rounded-sm border border-primary shadow focus-visible:outline-none focus-visible:ring-1 focus-visible:ring-ring disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground mt-[2px]"
      data-state="unchecked"
      role="checkbox"
      type="button"
      value="on"
    />
    <form
      action="/"
      class="w-full"
      method="post"
    >
      <input
        name="emailId"
        type="hidden"
        value="mock-id"
      />
       
      <button
        class="w-full cursor-pointer text-left data-[read=true]:opacity-50"
        data-read="false"
        name="_action"
        type="submit"
        value="setPreview"
      >
        <div
          class="mb-1 flex items-center justify-between "
        >
          <p
            class="w-max max-w-[180px] truncate text-sm font-bold"
          >
            mock-from
          </p>
          <p
            class="ml-auto w-max text-xs"
          >
            13:07
          </p>
        </div>
        <p
          class="mb-1 max-w-[250px] truncate text-sm"
        >
          mock-subject
        </p>
        <p
          class="line-clamp-2 max-w-[300px] text-xs"
        >
          mock-text
        </p>
      </button>
    </form>
  </div>
</div>
`;
